[AutoDiff] Canonicalize SIL type for JVP/VJP methods.

dan-zheng opened the pull request:
Canonicalize JVP/VJP type calculation for methods so that JVP/VJP methods return
a linear map that take/return self's tangent/cotangent last, instead of first.

This matches the type calculation logic for top-level functions.

Changes:
- Remove method-specific logic from
  `SILFunctionType::getAutoDiffAssociatedFunctionType`.
  Handle `self` like a normal differentiation parameter.
- Thunk user-defined JVP/VJP methods, reordering the position of self's
  tangent/cotangent in returned linear maps.
  - Relevant code:
    `SILGenModule::getOrCreateAutoDiffAssociatedFunctionReorderingThunk` and
    `SILGenFunction::getOrCreateAutoDiffLinearMapReorderingThunk`.
- Change type computation for JVP/VJP method protocol witnesses.
  - Relevant code: `getSILFunctionType` in SILFunctionType.cpp and
    `SILGenFunction::emitProtocolWitness`.

Move some functions to a common location:
- `AnyFunctionType::getAutoDiffOriginalFunctionType`
- `autodiff::getAutoDiffFunctionLinkage`

---

This is a step towards refactoring the differentiation transform (iterative `autodiff_function` worklist redesign), which will unblock control-flow support.

dan-zheng committed:
[AutoDiff] Canonicalize SIL type for JVP/VJP methods.

Canonicalize JVP/VJP type calculation for methods so that JVP/VJP methods return
a linear map that take/return self's tangent/cotangent last, instead of first.

This matches the type calculation logic for top-level functions.

Changes:
- Remove method-specific logic from
  `SILFunctionType::getAutoDiffAssociatedFunctionType`.
  Handle `self` like a normal differentiation parameter.
- Thunk user-defined JVP/VJP methods, reordering the position of self's
  tangent/cotangent in returned linear maps.
  - Relevant code:
    `SILGenModule::getOrCreateAutoDiffAssociatedFunctionReorderingThunk` and
    `SILGenFunction::getOrCreateAutoDiffLinearMapReorderingThunk`.
- Change type computation for JVP/VJP method protocol witnesses.
  - Relevant code: `getSILFunctionType` in SILFunctionType.cpp and
    `SILGenFunction::emitProtocolWitness`.

Move some functions to a common location:
- `AnyFunctionType::getAutoDiffOriginalFunctionType`
- `autodiff::getAutoDiffFunctionLinkage`

dan-zheng commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -84,7 +84,11 @@
 #include "SILGen.h"
 #include "SILGenFunction.h"
+// SWIFT_ENABLE_TENSORFLOW
+#include "SILGenFunctionBuilder.h"
```
It's possible to avoid introducing these dependencies by moving the "thunk function creation" logic to SILGenThunk.cpp.

This is done for reabstraction thunks:
* Thunk `SILFunction ` creation is done in SILGenThunk.cpp in `SILGenModule::getOrCreateReabstractionThunk`. SILGenThunk.cpp has the appropriate headers.
* Thunk body creation is done here in SILGenPoly.cpp.

Consider refactoring at some point.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4179,54 +4143,9 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-    // If the applied adjoint returns the adjoint of the original self
-    // parameter, then it returns it first. Accumulate adjoint for the original
-    // self parameter.
+    // Accumulate adjoints for original parameters.
```
```suggestion
    // Accumulate adjoints for original differentiation parameters.
```

dan-zheng commented:
Triggering tests to see if anything fails. More cleanup/doc comments coming up later.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -3953,6 +4210,61 @@ void SILGenFunction::emitProtocolWitness(AbstractionPattern reqtOrigTy,
+    bool isWrtSelf = indices.isWrtParameter(selfParamIndex);
+    if (isWrtSelf && indices.parameters->getNumIndices() > 1) {
+      auto callerFnType = F.getLoweredFunctionType();
+      auto assocFnType = F.mapTypeIntoContext(witnessFnRef->getType()).castTo<SILFunctionType>();
```
Fit to 80 columns and add doc comments.

dan-zheng committed:
Clean up, fix test.

- Fix test/Serialization/differentiable_attr.swift.
  - Use `CHECK` and `CHECK-NEXT` instead of `CHECK-DAG`.
- Remove dead code from `SILGenFunction::emitProtocolWitness`.
- Gardening.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -465,6 +466,10 @@ struct SILAutoDiffIndices {
 
   bool operator==(const SILAutoDiffIndices &other) const;
 
+  bool operator!=(const SILAutoDiffIndices &other) const {
```
I don't see any use in this file.

dan-zheng replied:
I used `!=` on occasion while debugging this and other patches. I'd like to keep it, if that's fine.

dan-zheng commented on lib/AST/Type.cpp:
```diff
@@ -4566,6 +4566,41 @@ AnyFunctionType *AnyFunctionType::getAutoDiffAssociatedFunctionType(
+AnyFunctionType *
+AnyFunctionType::getAutoDiffOriginalFunctionType() {
+  // Unwrap curry levels.
+  SmallVector<AnyFunctionType *, 2> curryLevels;
```
I see, that makes sense. Let's fix both `getAutoDiffAssociatedFunctionType` and `getAutoDiffOriginalFunctionType` in a separate patch.

dan-zheng replied:
Actually, let me commit the fix in this patch, since it's likely short.

Is it possible to do better than "always unwrap two parameter lists"?
```cpp
  auto *currentLevel = this;
  for (unsigned i : range(2)) {
    (void)i;
    if (currentLevel == nullptr)
      break;
    curryLevels.push_back(currentLevel);
    currentLevel = currentLevel->getResult()->getAs<AnyFunctionType>();
  }
```

I wonder how to detect "has self parameter" from an `AnyFunctionType`. Methods like `isInstanceMember` are defined on `FuncDecl` and aren't accessible.

dan-zheng replied:
Committed code above in https://github.com/apple/swift/pull/24775/commits/f8f4e89d446949b6abcc901857fbf5b5fecbdf38.

rxwei replied:
It's `getExtInfo().hasSelfParam()`.

dan-zheng replied:
`getExtInfo().hasSelfParam()` actually returns false for methods, for some reason:
```diff
diff --git a/lib/Sema/TypeCheckAttr.cpp b/lib/Sema/TypeCheckAttr.cpp
index a1509028f7..1edf502195 100644
--- a/lib/Sema/TypeCheckAttr.cpp
+++ b/lib/Sema/TypeCheckAttr.cpp
@@ -2917,8 +2917,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   }

   TC.resolveDeclSignature(original);
+  auto extInfo = original->getInterfaceType()->castTo<AnyFunctionType>()->getExtInfo();
+  llvm::errs() << "ORIGINAL FUNCTION '" << original->getNameStr() << "', HAS SELF PARAM? " << extInfo.hasSelfParam() << "\n";
+  original->getInterfaceType()->dump();
   auto *originalFnTy = original->getInterfaceType()->eraseDynamicSelfType()
       ->castTo<AnyFunctionType>();
   bool isMethod = original->hasImplicitSelfDecl();
```
```test.swift
struct S {
  @differentiable(vjp: vjpFoo)
  func foo(_ x: Float) -> Float {
    return x
  }

  func vjpFoo(_ x: Float) -> (Float, (Float) -> Float) {
    return (x, { $0 })
  }
}
```
```
$ swift test.swift
ORIGINAL FUNCTION 'foo', HAS SELF PARAM? 0
(function_type escaping
  (input=function_params num_params=1
    (param
      (struct_type decl=attr.(file).S@attr.swift:1:8)))
  (output=function_type escaping
    (input=function_params num_params=1
      (param
        (struct_type decl=Swift.(file).Float)))
    (output=struct_type decl=Swift.(file).Float)))
```

I'll keep the "unwrap two parameter lists if possible" code and not pursue this further.

dan-zheng committed:
Address review comments.

dan-zheng commented on include/swift/AST/AutoDiff.h:
```diff
@@ -21,6 +21,7 @@
 #include "ASTContext.h"
 #include "llvm/ADT/SmallBitVector.h"
 #include "swift/Basic/Range.h"
+#include "swift/SIL/SILLinkage.h"
```
Done in https://github.com/apple/swift/pull/24775/commits/f8f4e89d446949b6abcc901857fbf5b5fecbdf38.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

marcrasi reviewed (commented):
I have not read through all the type calculation and thunking in detail yet, because I have a high-level comment that might simplify some of it.

Can you make the AST type of the associated functions have self last? i.e. make AnyFunctionType::getAutoDiffAssociatedFunctionType return types with self last?

I haven't thought this through carefully, but I think that if you did that:
* This order wouldn't be exposed to the user anywhere, because the user never gets direct access to that AST type. (Except: You might have to make a small ordering adjustment in TypeCheckAttr).
* This would eliminate the reordering logic in `SILFunctionType.cpp`, because the order would already be correct.
* This would eliminate the reordering logic in `SILGenFuncion::emitProtocolWitness`, because the AST types that get there would already be in the right order.
* Also, this would be overall safer and less likely to break in the future, because every internal layer would have the same order. Only the very top layer that's exposed to the user would have a different order.

dan-zheng committed:
Address review comments.

- Add doc comments to `getAutoDiffFunctionLinkage`.
- Optimize "curry level unwrapping" in
  `AnyFunctionType::getAutoDiffAssociatedFunctionType`.

dan-zheng commented on include/swift/AST/AutoDiff.h:
```diff
@@ -574,6 +579,10 @@ bool getBuiltinAutoDiffApplyConfig(StringRef operationName,
                                    unsigned &arity, unsigned &order,
                                    bool &rethrows);
+
+SILLinkage getAutoDiffFunctionLinkage(SILLinkage originalLinkage,
```
Great suggestion, thanks! Documentation clarified in https://github.com/apple/swift/pull/24775/commits/d6ad9f244408c92e76e511145d81d2e7c8c9ace4.

dan-zheng committed:
Move self-parameter-reordering type computation from SIL to AST.

Address review feedback from @marcrasi.

Computing self-parameter-reordered JVP/VJP types on AST function types
instead of SIL function types is safer and simpler.

dan-zheng commented:
> Can you make the AST type of the associated functions have self last? i.e. make AnyFunctionType::getAutoDiffAssociatedFunctionType return types with self last?

Good idea! We discussed in person and I discussed with @rxwei - we agree that computing self-parameter-reordered JVP/VJP types on AST function types instead of SIL function types is safer and simpler.

Done in https://github.com/apple/swift/pull/24775/commits/063f4fd2b14fed4c9a380cee3ff89c8fad697dc7, removing 100+ lines from SILGen and adding <10 lines to AST/Sema.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
